### PR TITLE
Fix blocking threads in the node - Bindings

### PR DESCRIPTION
**Pull request by jotare**
### Description
- Bindings now use the new cache abstractions
- NodeReader and NodeWriter will be removed soon (as they are no longer needed)

### How was this PR tested?
- Already existing tests

### CODE REVIEW

File: nucliadb_node/binding/src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use pyo3::create_exception;
+use pyo3::exceptions::PyException;
+
+// Base exception for all custom errors produced by this library
+create_exception!(nucliadb_node_binding, IndexNodeException, PyException);
+
+create_exception!(nucliadb_node_binding, LoadShardError, IndexNodeException);
+create_exception!(nucliadb_node_binding, ShardNotFound, IndexNodeException);
```
Comment (hermeGarcia) [inline nucliadb_node/binding/src/errors.rs:27]: Good one!

File: nucliadb_node/src/shards/unbounded_cache_provider/unbounded_writer.rs [hunk 2]
```diff
@@ -16,3 +16,126 @@
...
+            let entry = entry?;
+            let file_name = entry.file_name().to_str().unwrap().to_string();
+            let shard_path = entry.path();
+            match ShardWriter::open(file_name.clone(), &shard_path) {
+                Err(err) => error!("Loading shard {shard_path:?} from disk raised {err}"),
+                Ok(shard) => {
+                    debug!("Shard loaded: {shard_path:?}");
+                    cache.insert(file_name, Arc::new(shard));
+                }
+            }
```
Comment (hermeGarcia) [inline nucliadb_node/src/shards/unbounded_cache_provider/unbounded_writer.rs:106]: I think I would have used: 
```rust
            let shard = ShardWriter::open(file_name.clone(), &shard_path)
                .map_err(|err| err.context(format!("Loading {shard_path:?} from disk")))?;
            debug!("Shard loaded: {shard_path:?}");
            cache.insert(file_name, Arc::new(shard));
```

File: nucliadb_node/src/writer/grpc_driver.rs [hunk 3]
```diff
@@ -269,16 +275,19 @@ impl NodeWriter for NodeWriterGRPCDriver {
...
-        request: Request<ResourceId>,
+        request: Request<DeleteGraphNodes>,
     ) -> Result<Response<OpStatus>, Status> {
-        let resource = request.into_inner();
-        let shard_id = resource.shard_id.clone();
-        let shard = self.obtain_shard(&shard_id).await?;
+        let request = request.into_inner();
+        let shard_id = match request.shard_id {
+            Some(ref shard_id) => &shard_id.id,
+            None => return Err(tonic::Status::invalid_argument("Shard ID must be provided")),
```
Comment (hermeGarcia) [inline nucliadb_node/src/writer/grpc_driver.rs:285]: I think I would have gone for:
```rust
        let Some(shard_id) = request.shard_id.as_ref().map(|i| &i.id) else {
            return Err(tonic::Status::invalid_argument("Shard ID must be provided"));
        };
```

File: nucliadb_node/src/writer/grpc_driver.rs [hunk 4]
```diff
@@ -237,7 +234,16 @@ impl NodeWriter for NodeWriterGRPCDriver {
...
+            Err(error) => {
+                let status = OpStatus {
+                    status: op_status::Status::Error as i32,
+                    detail: error.to_string(),
+                    field_count: 0_u64,
+                    shard_id,
+                    ..Default::default()
+                };
+                Ok(tonic::Response::new(status))
+            }
```
Comment (hermeGarcia) [inline nucliadb_node/src/writer/grpc_driver.rs:246]: Maybe using one expression simplifies the code:
```rust
            Err(error) => Ok(tonic::Response::new(OpStatus {
                status: op_status::Status::Error as i32,
                detail: error.to_string(),
                field_count: 0_u64,
                shard_id,
                ..Default::default()
            })),
```

File: nucliadb_node/src/writer/grpc_driver.rs (hunk 3 above)
Comment (jotare) [reply nucliadb_node/src/writer/grpc_driver.rs:285]: This is code moved inside the file and the style is the same across all functions. I don't see much benefit changing it

File: nucliadb_node/src/writer/grpc_driver.rs (hunk 4 above)
Comment (jotare) [reply nucliadb_node/src/writer/grpc_driver.rs:246]: Same

**Review by hermeGarcia — APPROVED** [review]

**Pull request merged**